pkg/apiserver/proxy_test.go

```diff
@@ -34,6 +36,124 @@ import (
 	"k8s.io/kubernetes/pkg/api/rest"
 )
 
+func calculateContentLength(reqBody string, transferEncodings []string) int64 {
+	if len(transferEncodings) == 0 || (len(transferEncodings) == 1 && transferEncodings[0] == "identity") {
+		return int64(len(reqBody))
+	}
+	// RFC2616 section 4.4: The Content-Length header field MUST NOT be sent if these two lengths
+	// are different (i.e., if a Transfer-Encoding header field is present).
+	return int64(-1)
+}
+
+func shouldGzip(transferEncodings []string) bool {
+	for _, encoding := range transferEncodings {
+		if encoding == "gzip" {
+			return true
+		}
+	}
+	return false
+}
+
+func TestProxyRequestContentLengthAndTransferEncoding(t *testing.T) {
+	serverResponse := "got response"
+	table := []struct {
+		transferEncodings []string
+		contentEncoding   string
+		reqBody           string
+		reqNamespace      string
+	}{
+		{[]string{}, "", "", "default"},
+		{[]string{"identity"}, "", "question", "default"},
+		{[]string{"chunked"}, "", "question", "default"},
+		// RFC2616 section-3.6: Whenever a transfer-coding is applied to a message-body, the set of
+		// transfer-codings MUST include "chunked", unless the message is terminated by closing the
+		// connection.
+		{[]string{"chunked", "gzip"}, "gzip", "qqqqqqqqqquestion", "default"},
+	}
+
+	for _, item := range table {
+		downstreamServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			fmt.Println("CHAO: downstream got header: ", req.Header)
+			expectedContentLength := calculateContentLength(item.reqBody, item.transferEncodings)
+			if e, a := expectedContentLength, req.ContentLength; e != a {
+				t.Errorf("expected %v, got %v", e, a)
+			}
+			// We expect Content-Length header field be set
+			if expectedContentLength != -1 {
+				contentLengthInHeader, err := strconv.Atoi(req.Header.Get("Content-Length"))
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if e, a := expectedContentLength, int64(contentLengthInHeader); e != a {
+					t.Errorf("expected %v, got %v", e, a)
+				}
+			}
+
+			// The http library will strip the "Transfer-Encoding" field from request.Header.
```

Comment:
As the comment reads, Go library strips the "Transfer-Encoding" field. It will stores the value in request.TransferEncodings after some processing, the only possible value will be either "" or "chunked".

People say usually "gzip" will be a "Content-Encoding", so hopefully we won't get bit for this.

The discussion in go-nuts group: https://groups.google.com/forum/#!topic/golang-nuts/g7xJk09xBxU